game/video.cc

```diff
@@ -47,40 +47,46 @@ Video::~Video() {
-	// make ffmpeg here to get any exception in the current thread
-	auto ffmpeg = std::make_unique<VideoFFmpeg>(_videoFile, [this] (auto f) { this->push(std::move(f)); });
-	m_grabber = std::async(std::launch::async, [this, file = _videoFile, ffmpeg = std::move(ffmpeg)] {
-		int errors = 0;
-		std::unique_lock<std::mutex> l(m_mutex);
-		while (!m_quit) {
-			if (m_seek_asked) {
-				m_seek_asked = false;
-
-				auto seek_pos = m_readPosition;
-				// discard all outdated frame. To avoid races between clean and push, clean and push are done in this thread.
-				m_queue.clear();
-
-				UnlockGuard<decltype(l)> unlocked(l); // release lock during seek
-				ffmpeg->seek(seek_pos);
-				continue;
-			}
+	m_grabber = std::async(std::launch::async, [this, file = _videoFile] {
+		try {
+			auto ffmpeg = std::make_unique<VideoFFmpeg>(file, [this](auto f) { this->push(std::move(f)); });
+			int errors = 0;
+			std::unique_lock<std::mutex> l(m_mutex);
+			while (!m_quit) {
+				if (m_seek_asked) {
+					m_seek_asked = false;
+
+					auto seek_pos = m_readPosition;
+					// discard all outdated frame. To avoid races between clean and push, clean and push are done in this thread.
+					m_queue.clear();
+
+					UnlockGuard<decltype(l)> unlocked(l);  // release lock during seek
+					ffmpeg->seek(seek_pos);
+					continue;
+				}
 
-			try {
-				UnlockGuard<decltype(l)> unlocked(l); // release lock during possibly blocking ffmpeg stuff
-				ffmpeg->handleOneFrame();
-				errors = 0;
-			} catch (FFmpeg::Eof&) {
-				{
-					UnlockGuard<decltype(l)> unlocked(l); // release lock for possibly blocking calls
-					push(Bitmap()); // EOF marker
-					std::clog << "ffmpeg/debug: done loading " << file << std::endl;
+				try {
+					UnlockGuard<decltype(l)> unlocked(l);  // release lock during possibly blocking ffmpeg stuff
+					ffmpeg->handleOneFrame();
+					errors = 0;
+				} catch (FFmpeg::Eof&) {
+					{
+						UnlockGuard<decltype(l)> unlocked(l);  // release lock for possibly blocking calls
+						push(Bitmap());						   // EOF marker
+						std::clog << "ffmpeg/debug: done loading " << file << std::endl;
+					}
+					m_cond.wait(l, [this] { return m_quit || m_seek_asked; });
+				} catch (std::exception& e) {
+					UnlockGuard<decltype(l)> unlocked(l);  // release lock for possibly blocking calls
+					std::clog << "ffmpeg/error: " << file << ": " << e.what() << std::endl;
+					if (++errors > 2) {
+						std::clog << "ffmpeg/error: FFMPEG terminating due to multiple errors" << std::endl;
+						break;
+					}
 				}
-				m_cond.wait(l, [this]{ return m_quit || m_seek_asked; });
-			} catch (std::exception& e) {
-				UnlockGuard<decltype(l)> unlocked(l); // release lock for possibly blocking calls
-				std::clog << "ffmpeg/error: " << file << ": " << e.what() << std::endl;
-				if (++errors > 2) { std::clog << "ffmpeg/error: FFMPEG terminating due to multiple errors" << std::endl; break; }
 			}
+		} catch (std::exception& e) {
```

Comment:
maybe just keep the try/catch around VideoFFmpeg if it is the only place to throw; here it is quite hard to tell where the exception can come from.

Reply:
I'd do that but then the `ffmpeg` variable lives only inside that scope and cannot be used in later code. All the alternatives are somewhat ugly (like not using `auto` but manually defining the variable).

Reply:
ok